### PR TITLE
Changes for windows (memory map, memory modules, radare commands with pipe)

### PR DESCRIPTION
**Pull request by josediazfer**
FIX: All the memory regions of the process are shown (previously only the modules were shown).
FIX: Before only the 64-bit modules were shown, adding the flag TH32CS_SNAPMODULE32 also shows the 32bits.
FIX: In Windows 7 the stat () function fails if the temporary path ends with slash.
The r_core_cmd_pipe () function is implemented.

### CODE REVIEW

File: libr/core/cmd.c [hunk 1]
```diff
@@ -1252,6 +1252,80 @@ static int cmd_system(void *data, const char *input) {
 	return ret;
 }
 
+#if __WINDOWS__ && !__CYGWIN__
+static void r_w32_cmd_pipe(RCore *core, char *radare_cmd, char *shell_cmd) {
+	STARTUPINFO si;
+        PROCESS_INFORMATION pi;
```
Comment (radare) [inline libr/core/cmd.c]: Use tabs to indent

File: libr/core/cmd.c [hunk 2]
```diff
@@ -1252,6 +1252,80 @@ static int cmd_system(void *data, const char *input) {
...
+	SECURITY_ATTRIBUTES sa;
+	HANDLE pipe[2];
+	int fd_out = -1, cons_out = -1;
+	char *_shell_cmd;
+
+	sa.nLength = sizeof (SECURITY_ATTRIBUTES); 
+	sa.bInheritHandle = TRUE; 
+	sa.lpSecurityDescriptor = NULL; 
+	memset (&si, 0, sizeof (si));
```
Comment (radare) [inline libr/core/cmd.c]: Initialize structs with {0} in the definition not with memset

File: libr/core/cmd.c [hunk 3]
```diff
@@ -1252,6 +1252,80 @@ static int cmd_system(void *data, const char *input) {
...
+	memset (&pi, 0, sizeof (pi));
+	memset (pipe, 0, sizeof (pipe));
+   	if (!CreatePipe(&pipe[0], &pipe[1], &sa, 0)) {
+		r_sys_perror ("r_w32_cmd_pipe/CreatePipe");
+		goto err_r_w32_cmd_pipe;
+	}	
+	if (!SetHandleInformation(pipe[1], HANDLE_FLAG_INHERIT, 0)) {
```
Comment (radare) [inline libr/core/cmd.c]: Add slace before every (

File: libr/core/cmd.c [hunk 4]
```diff
@@ -1252,6 +1252,80 @@ static int cmd_system(void *data, const char *input) {
...
+	_shell_cmd = shell_cmd;
+	while (*_shell_cmd && isspace(*_shell_cmd)) {
+		_shell_cmd++;
+	}
+	// exec windows process
+	if (!CreateProcess (NULL, _shell_cmd, NULL, NULL, TRUE, 0, NULL, NULL, &si, &pi)) {
+		r_sys_perror ("r_w32_cmd_pipe/CreateProcess");
+		goto err_r_w32_cmd_pipe;
+	}
+	fd_out = _open_osfhandle((intptr_t)pipe[1], _O_WRONLY|_O_TEXT);
```
Comment (radare) [inline libr/core/cmd.c]: Missing space

File: libr/core/cmd.c [hunk 5]
```diff
@@ -1252,6 +1252,80 @@ static int cmd_system(void *data, const char *input) {
...
+	if (fd_out == -1) {
+		perror("_open_osfhandle");
+		goto err_r_w32_cmd_pipe;
+	}
+	cons_out = dup(1);
+	dup2(fd_out, 1);
```
Comment (radare) [inline libr/core/cmd.c]: Dup2 on windows?

File: libr/core/cmd.c [hunk 6]
```diff
@@ -1252,6 +1252,80 @@ static int cmd_system(void *data, const char *input) {
...
+	close(fd_out);
+	fd_out = -1;
+	WaitForSingleObject(pi.hProcess, INFINITE);
+err_r_w32_cmd_pipe:
+	if (pi.hProcess != NULL) {
+		CloseHandle(pi.hProcess);
+	}
+	if (pi.hThread != NULL) {
+		CloseHandle(pi.hThread);
+	}
```
Comment (radare) [inline libr/core/cmd.c]: Handles are not void*

File: libr/debug/p/native/maps/windows.c [hunk 7]
```diff
@@ -1,200 +1,291 @@
...
+typedef struct {
+	RDebugMap *map;
+	IMAGE_SECTION_HEADER *sect_hdr;
+	int sect_count;
+} RWinModInfo;
+
+static char *get_map_type(MEMORY_BASIC_INFORMATION *mbi) {
+	char *type;
+	switch (mbi->Type) {
+		case MEM_IMAGE:
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Cases are indented at the same level of the switch

File: libr/debug/p/native/maps/windows.c [hunk 8]
```diff
@@ -1,200 +1,291 @@
...
-
-	hModuleSnap = CreateToolhelp32Snapshot (TH32CS_SNAPMODULE, pid);
-	if(!hModuleSnap ) {
-		r_sys_perror ("w32_dbg_maps/CreateToolhelp32Snapshot");
-		CloseHandle( hModuleSnap );
-		return NULL;
+	int perm;
+	char *map_name, *map_type;
+
+	map_type = get_map_type (mbi);
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Define and assign i  the same line

File: libr/debug/p/native/maps/windows.c [hunk 9]
```diff
@@ -1,200 +1,291 @@
...
+	map_name = (char *)malloc(strlen(name) +
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: CHeck for nulls, add spaces and use calloc

File: libr/debug/p/native/maps/windows.c [hunk 10]
```diff
@@ -1,200 +1,291 @@
...
-		CloseHandle( hModuleSnap );
-		return NULL;
+	int perm;
+	char *map_name, *map_type;
+
+	map_type = get_map_type (mbi);
+	map_name = (char *)malloc(strlen(name) + 
+				  strlen(map_type) +
+				  50);
+	if (map_name == NULL) {
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Use ! Instead of NULL

File: libr/debug/p/native/maps/windows.c [hunk 11]
```diff
@@ -1,200 +1,291 @@
...
+		perror("malloc name add_map()");
+		goto err_add_map;
 	}
-	me32.dwSize = sizeof(MODULEENTRY32);
-	if (!Module32First (hModuleSnap, &me32))	{
-		CloseHandle (hModuleSnap);
-		return NULL;
+	switch (mbi->Protect) {
+		case PAGE_EXECUTE:
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Wrong indent

File: libr/debug/p/native/maps/windows.c [hunk 12]
```diff
@@ -1,200 +1,291 @@
...
-			mapname = (char *)malloc(MAX_PATH);
-			if (!mapname) continue;
-			for (i = 0; i < NumSections; i++) {
-				if (SectionHeader->Misc.VirtualSize <= 0)
-					continue;
-				sprintf(mapname,"%s | %s", me32.szModule, SectionHeader->Name);
-				mr = r_debug_map_new (mapname, (ut64)(size_t)(SectionHeader->VirtualAddress + me32.modBaseAddr), (ut64)(size_t)(SectionHeader->VirtualAddress + me32.modBaseAddr + SectionHeader->Misc.VirtualSize), SectionHeader->Characteristics, 0);
-				if (mr != NULL) r_list_append (list, mr);
-				SectionHeader++;
+	sprintf(map_name, "%-8s %s", map_type, name);
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Overflow here. Use rstrnewf

File: libr/debug/p/native/maps/windows.c [hunk 13]
```diff
@@ -1,200 +1,291 @@
...
+	mr = r_debug_map_new (map_name,
+			     addr,
+			     addr + len, perm, mbi->Type == MEM_PRIVATE);
+	if (mr)  {
+		r_list_append (list, mr);
+	}
+err_add_map:
+	if (map_name) {
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Unnecessaryif

File: libr/debug/p/native/maps/windows.c [hunk 14]
```diff
@@ -1,200 +1,291 @@
...
+		free (map_name);
+	}
+	return mr;
+}
+
+static RDebugMap *add_map_reg(RList *list, const char *name, MEMORY_BASIC_INFORMATION *mbi) {
+	return add_map(list, name, (ut64)mbi->BaseAddress, (ut64)mbi->RegionSize, mbi);
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Missing space

File: libr/debug/p/native/maps/windows.c [hunk 15]
```diff
@@ -1,200 +1,291 @@
...
+static RList *w32_dbg_modules(RDebug *dbg) {
+        HANDLE h_mod_snap;
+        MODULEENTRY32 me32;
+        RDebugMap *mr;
+        int pid = dbg->pid;
+        RList *list = r_list_new ();
+
+        h_mod_snap = CreateToolhelp32Snapshot( TH32CS_SNAPMODULE | TH32CS_SNAPMODULE32, pid );
+        if (!h_mod_snap) {
+                r_sys_perror ("w32_dbg_modules/CreateToolhelp32Snapshot");
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Use tabs

File: libr/debug/p/native/maps/windows.c [hunk 16]
```diff
@@ -1,200 +1,291 @@
...
+        }
+        do {
+                ut64 baddr = (ut64)(size_t)me32.modBaseAddr;
+                mr = r_debug_map_new (me32.szModule, baddr, baddr + me32.modBaseSize, 0, 0);
+                if (mr != NULL) {
+                        mr->file = strdup (me32.szExePath);
+                        if (mr->file != NULL)
+                                r_list_append (list, mr);
+                }
+        } while(Module32Next (h_mod_snap, &me32));
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Missing spaces

File: libr/debug/p/native/maps/windows.c [hunk 17]
```diff
@@ -1,200 +1,291 @@
...
+	len = 0;
+	sect_hdr = NULL;
+	mod_inf_fill = -1;
+	ReadProcessMemory (h_proc, (LPCVOID)map->addr, (LPVOID)pe_hdr, sizeof(pe_hdr), &len);
+	if (len == sizeof (pe_hdr) && is_pe_hdr (pe_hdr)) {
+		dos_hdr = (IMAGE_DOS_HEADER *)pe_hdr;
+		if (!dos_hdr)
+			return -1;
+		nt_hdrs = (IMAGE_NT_HEADERS *)((char *)dos_hdr + dos_hdr->e_lfanew);
+		if (!nt_hdrs)
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Add braces

File: libr/debug/p/native/maps/windows.c [hunk 18]
```diff
@@ -1,200 +1,291 @@
...
+			return -1;
+		if (nt_hdrs->FileHeader.Machine == 0x014c) { // check for x32 pefile
+			nt_hdrs32 = (IMAGE_NT_HEADERS32 *)((char *)dos_hdr + dos_hdr->e_lfanew);
+			mod->sect_count = nt_hdrs32->FileHeader.NumberOfSections;
+			sect_hdr = (IMAGE_SECTION_HEADER *)((char *)nt_hdrs32 + sizeof (IMAGE_NT_HEADERS32));
+		} else {
+			mod->sect_count = nt_hdrs->FileHeader.NumberOfSections;
+			sect_hdr = (IMAGE_SECTION_HEADER *)((char *)nt_hdrs + sizeof (IMAGE_NT_HEADERS));
+		}
+		if (sect_hdr != NULL) {
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: !=NULL is not necessary

File: libr/debug/p/native/maps/windows.c [hunk 19]
```diff
@@ -1,200 +1,291 @@
...
-	if (!gmi) {
-		eprintf ("w32dbg: no gmi\n");
-		return 0;
 	}
-	if (!gmbn) {
-		eprintf ("w32dbg: no gmn\n");
-		return 0;
+err_set_mod_info:
+	if (mod_inf_fill == -1) {
+		if (mod->sect_hdr) {
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Unnecessary if

File: libr/debug/p/native/maps/windows.c [hunk 20]
```diff
@@ -1,200 +1,291 @@
...
+			free (mod->sect_hdr);
+		}	
 	}
+	return mod_inf_fill;
+}
+
+static void proc_mem_img(HANDLE h_proc, RList *map_list, RList *mod_list, RWinModInfo *mod, SYSTEM_INFO *si, MEMORY_BASIC_INFORMATION *mbi) {
+	ut64 addr, len;
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Define and assign in the same line. We do c99 now

File: libr/debug/p/native/maps/windows.c [hunk 21]
```diff
@@ -1,200 +1,291 @@
...
+
+	addr = (ut64)mbi->BaseAddress;
+	len = (ut64)mbi->RegionSize;
+	if (mod->map == NULL || addr < mod->map->addr || (addr + len) > mod->map->addr_end) {
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Use ! Instead of == NULL

File: libr/debug/p/native/maps/windows.c [hunk 22]
```diff
@@ -1,200 +1,291 @@
...
 
-#if !__MINGW64__	// TODO: Fix this , for win64 cant walk over all process memory, use psapi.dll to get modules
-	for (page=(LPBYTE)SysInfo.lpMinimumApplicationAddress;
-			page<(LPBYTE)SysInfo.lpMaximumApplicationAddress;) {
-		if (!VirtualQueryEx (WIN32_PI (hProcess), page, &mbi, sizeof (mbi)))  {
-	//		eprintf ("VirtualQueryEx ERROR, address = 0x%08X\n", page);
-			page += SysInfo.dwPageSize;
-			continue;
-			//return NULL;
+		if (mod->sect_hdr != NULL) {
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: remove that if

File: libr/debug/p/native/maps/windows.c [hunk 23]
```diff
@@ -1,200 +1,291 @@
...
-			if (ret_len == sizeof (PeHeader) && CheckValidPE (PeHeader)) {
-				dos_header = (IMAGE_DOS_HEADER *)PeHeader;
-				if (!dos_header)
-					break;
-				nt_headers = (IMAGE_NT_HEADERS *)((char *)dos_header
-						+ dos_header->e_lfanew);
-				if (!nt_headers) {
-					// skip before failing
-					break;
+		memset (mod, 0, sizeof(RWinModInfo));
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Missing space before (

File: libr/debug/p/native/maps/windows.c [hunk 24]
```diff
@@ -1,200 +1,291 @@
...
-				eprintf ("Invalid read\n");
-				return NULL;
+				free (map_name);
+				sect_count++;
 			}
+		}
+		if (sect_count == 0) {
+			add_map_reg (map_list, mod->map->name, mbi);
+		}
+	} else {
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Swap condition and remove ==NULL

Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Add braces {}

File: libr/debug/p/native/maps/windows.c [hunk 25]
```diff
@@ -1,200 +1,291 @@
...
-				return NULL;
+err_proc_mem_img:
+	;
+}
+
+static void proc_mem_map(HANDLE h_proc, RList *map_list, MEMORY_BASIC_INFORMATION *mbi) {
+	char f_name[MAX_PATH + 1];
+	DWORD len;
+
+	len = w32_GetMappedFileName (h_proc, mbi->BaseAddress, f_name, MAX_PATH);
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Define and assign in the same line

File: libr/debug/p/native/maps/windows.c [hunk 26]
```diff
@@ -1,200 +1,291 @@
...
+	HANDLE h_proc;
+	RList *map_list, *mod_list;
+	RWinModInfo mod_inf;
+
+	map_list = r_list_new ();
+	memset(&mod_inf, 0, sizeof(mod_inf));
+	memset (&si, 0, sizeof (si));
+	GetSystemInfo (&si);
+	h_proc = w32_openprocess (PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, dbg->pid);
+	if (h_proc == NULL) {
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Use !

File: libr/debug/p/native/maps/windows.c [hunk 27]
```diff
@@ -1,200 +1,291 @@
...
+	}
+	cur_addr = si.lpMinimumApplicationAddress;
+	/* get process modules list */
+	mod_list = w32_dbg_modules(dbg);
+	/* process memory map */
+	while (cur_addr < si.lpMaximumApplicationAddress && 
+	       VirtualQueryEx (h_proc, cur_addr, &mbi, sizeof (mbi)) != 0) {
+	       	if (mbi.State != MEM_FREE) {
+			switch (mbi.Type) {
+				case MEM_IMAGE:
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Wrong inde t

File: libr/debug/p/native/maps/windows.c [hunk 28]
```diff
@@ -1,200 +1,291 @@
...
+	   	cur_addr = (LPVOID)((ut64)mbi.BaseAddress + mbi.RegionSize);
+	}
+err_w32_dbg_maps:
+	if (mod_inf.sect_hdr != NULL) {
+		free (mod_inf.sect_hdr);
 	}
-#endif
-	return list;
-*/
+	r_list_free(mod_list);
```
Comment (radare) [inline libr/debug/p/native/maps/windows.c]: Missing space

File: libr/util/file.c [hunk 29]
```diff
@@ -889,6 +889,13 @@ R_API char *r_file_tmpdir() {
 		}
 		path = strdup (tmpdir);
 	}
+	// Windows 7, stat() function fail if tmpdir ends with '\\'
+	if (path != NULL) {
+		int path_len = strlen(path);
+		if (path_len > 0 && path[path_len - 1] == '\\')
```
Comment (radare) [inline libr/util/file.c]: Missing braces

File: libr/util/file.c [hunk 30]
```diff
@@ -889,6 +889,13 @@ R_API char *r_file_tmpdir() {
 		}
 		path = strdup (tmpdir);
 	}
+	// Windows 7, stat() function fail if tmpdir ends with '\\'
+	if (path != NULL) {
+		int path_len = strlen(path);
+		if (path_len > 0 && path[path_len - 1] == '\\') 
+			path[path_len - 1] = (char)NULL;
```
Comment (radare) [inline libr/util/file.c]: Just use 0 and avoid the unnecessary cast and null

**Comment by xarkes** [comment]
It should fix https://github.com/radare/radare2/issues/7947 if this thing about stat() and extra backslash is true.

**Comment by radare** [comment]
Cc @skuater

**Pull request merged**